### PR TITLE
#459 Upgrade to Akka 2.5

### PR DESCRIPTION
**Pull request by anilgursel**
Thanks for your pull request.  Please review the following guidelines.

- [x] Title includes issue id.
- [x] Description of the change added.
- [x] Commits are squashed.
- [x] Tests added.
- [ ] Documentation added/updated.
- [x] Also please review [CONTRIBUTING.md](https://github.com/paypal/squbs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: squbs-ext/src/main/scala/akka/http/org/squbs/util/JavaConverters.scala [hunk 1]
```diff
@@ -57,14 +57,20 @@ object JavaConverters {
...
   }
 
-  private def adaptTupleFlow[T, Mat](scalaFlow: Flow[(HttpRequest, T), (Try[HttpResponse], T), Mat]):
-  js.Flow[Pair[jm.HttpRequest, T], Pair[Try[jm.HttpResponse], T], Mat] = {
+  private def adaptTupleFlow[T](scalaFlow: Flow[(HttpRequest, T), (Try[HttpResponse], T), HostConnectionPool]):
+  js.Flow[Pair[jm.HttpRequest, T], Pair[Try[jm.HttpResponse], T], jd.HostConnectionPool] = {
     implicit val _ = JavaMapping.identity[T]
+    implicit object HostConnectionPoolMapping extends JavaMapping[jd.HostConnectionPool, HostConnectionPool] {
+      def toScala(javaObject: jd.HostConnectionPool): HostConnectionPool =
+        throw new UnsupportedOperationException("jd.HostConnectionPool cannot be converted to Scala")
```
Comment (anilgursel) [inline squbs-ext/src/main/scala/akka/http/org/squbs/util/JavaConverters.scala]: This is something I need to still work on.

File: squbs-httpclient/build.sbt [hunk 2]
```diff
@@ -7,6 +7,7 @@ javaOptions in Test += "-Xmx512m"
 libraryDependencies ++= Seq(
   "com.typesafe.akka"         %% "akka-actor"                   % akkaV,
   "com.typesafe.akka"         %% "akka-slf4j"                   % akkaV,
+  "com.typesafe.akka"         %% "akka-stream"                  % akkaV,
```
Comment (anilgursel) [inline squbs-httpclient/build.sbt]: Please see https://github.com/akka/akka-http/issues/821

File: squbs-ext/src/main/scala/org/squbs/streams/Deduplicate.scala [hunk 3]
```diff
@@ -49,10 +49,14 @@ object Deduplicate {
   */
 final class Deduplicate[T, U](key: T => U, duplicateCount: Long = Long.MaxValue,
                               registry: java.util.Map[U, MutableLong] = new java.util.HashMap[U, MutableLong]())
-  extends SimpleLinearGraphStage[T] {
+  extends GraphStage[FlowShape[T, T]] {
```
Comment (anilgursel) [inline squbs-ext/src/main/scala/org/squbs/streams/Deduplicate.scala]: It is `@InternalApi private[akka]`: https://github.com/akka/akka/blob/v2.5.2/akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala#L52

File: squbs-pattern/src/main/scala/org/squbs/pattern/orchestration/japi/AbstractOrchestrator.scala [hunk 4]
```diff
@@ -17,17 +17,15 @@ package org.squbs.pattern.orchestration.japi
...
-
-//import akka.pattern.ask
 import akka.util.Timeout
 import org.squbs.pattern.orchestration.Orchestrator
 
 import scala.concurrent.Future
 import scala.util.{Failure, Success, Try}
 
-abstract class AbstractOrchestrator extends AbstractActor with Orchestrator {
+abstract class AbstractOrchestrator extends Actor with Orchestrator {
```
Comment (anilgursel) [inline squbs-pattern/src/main/scala/org/squbs/pattern/orchestration/japi/AbstractOrchestrator.scala]: Please see https://github.com/paypal/squbs/issues/470.

File: squbs-ext/src/main/scala/akka/http/org/squbs/util/JavaConverters.scala (hunk 1 above)
Comment (akara) [inline squbs-ext/src/main/scala/akka/http/org/squbs/util/JavaConverters.scala]: OK. Lets walk through it when you get it in better shape.

File: squbs-pattern/src/test/java/org/squbs/pattern/orchestration/japi/AskOrchestrator.java [hunk 5]
```diff
@@ -38,9 +38,9 @@ public AskOrchestrator(FiniteDuration orchDelay, Timeout askTimeout) {
         this.orchDelay = orchDelay;
         this.askTimeout = askTimeout;
 
-        expectOnce(ReceiveBuilder.match(TestRequest.class,
-                testRequest -> orchestrate(testRequest, sender())
-        ).build());
+        expectOnce(ReceiveBuilder.create().match(TestRequest.class,
+                testRequest -> orchestrate(testRequest, sender())).build().onMessage());
```
Comment (akara) [inline squbs-pattern/src/test/java/org/squbs/pattern/orchestration/japi/AskOrchestrator.java]: I'm good with this, temporarily. Ideally we should see a block like this for Java: 

```java
expectOnce(r -> r.match(TestRequest.class, testRequest -> orchestrate(...)));
```

File: squbs-pattern/src/test/java/org/squbs/pattern/orchestration/japi/SelectionOrchestrator.java [hunk 6]
```diff
@@ -30,10 +30,10 @@
 public class SelectionOrchestrator extends AbstractOrchestrator {
 
     public SelectionOrchestrator() {
-        getContext().actorOf(Props.create(ServiceEmulator.class), "ServiceEmulator");
-        expectOnce(ReceiveBuilder.match(TestRequest.class,
+        context().actorOf(Props.create(ServiceEmulator.class), "ServiceEmulator");
+        expectOnce(ReceiveBuilder.create().match(TestRequest.class,
```
Comment (akara) [inline squbs-pattern/src/test/java/org/squbs/pattern/orchestration/japi/SelectionOrchestrator.java]: Ditto

File: squbs-testkit/src/test/java/org/squbs/testkit/TestActorJ.java [hunk 7]
```diff
@@ -16,12 +16,19 @@
...
+import akka.actor.AbstractActor;
 
-public class TestActorJ extends UntypedActor {
+public class TestActorJ extends AbstractActor {
     public void onReceive(Object msg) {
         if ("Ping".equals(msg)) {
             getSender().tell("Pong", getSelf());
         }
     }
+
```
Comment (akara) [inline squbs-testkit/src/test/java/org/squbs/testkit/TestActorJ.java]: Do we have to remove the `onReceive` from this code?

File: squbs-pattern/src/test/java/org/squbs/pattern/orchestration/japi/AskOrchestrator.java (hunk 5 above)
Comment (anilgursel) [inline squbs-pattern/src/test/java/org/squbs/pattern/orchestration/japi/AskOrchestrator.java]: Shouldn't it actually be

```java
expectOnce(receiveBuilder().match(TestRequest.class, testRequest -> orchestrate(...)));
```

File: squbs-ext/src/main/scala/akka/http/org/squbs/util/JavaConverters.scala (hunk 1 above)
Comment (anilgursel) [inline squbs-ext/src/main/scala/akka/http/org/squbs/util/JavaConverters.scala]: I will look into this.  If you have any ideas, let me know.

File: squbs-testkit/src/test/java/org/squbs/testkit/TestActorJ.java (hunk 7 above)
Comment (anilgursel) [inline squbs-testkit/src/test/java/org/squbs/testkit/TestActorJ.java]: You're right.  We should remove.

**Pull request merged**